### DISCUSSION TITLE
What are the input parameters? @aws-sdk/client-bedrock-agent-runtime

### DISCUSSION THREAD

**Discussion by terrarishabh** [aws/aws-sdk-js-v3#6289] [Q&A]
I want to use `@aws-sdk/client-bedrock-agent-runtime` to interact with my bedrock agent, is this the correct thing to use?

Also, what do i need to pass in:
```
const params = {
  /** input parameters */
};
```

**Comment by kimnamu** [comment]
Thanks for including the exact README snippet; that `const params = { /** input parameters */ }` line is the generic template every client README shares, so it never says which fields InvokeAgent needs.

**Short version:** yes, `@aws-sdk/client-bedrock-agent-runtime` with `InvokeAgentCommand` is the package for talking to an agent you already have (creating and preparing agents lives in `@aws-sdk/client-bedrock-agent`). Only three fields are required, all path parameters, and `inputText` carries the user's message. The response's `completion` is an async iterable of events, with the agent's text in `event.chunk.bytes`.

| Field | Required | Where it goes |
|---|---|---|
| `agentId` | yes | URL path |
| `agentAliasId` | yes | URL path |
| `sessionId` | yes | URL path; any string you choose, reuse it to continue the same conversation |
| `inputText` | no (ignored when `sessionState.returnControlInvocationResults` is set) | JSON body; the prompt |
| `enableTrace`, `endSession`, `sessionState`, `memoryId`, ... | no | JSON body |

```ts
import { BedrockAgentRuntimeClient, InvokeAgentCommand } from "@aws-sdk/client-bedrock-agent-runtime";

const client = new BedrockAgentRuntimeClient({ region: "us-east-1" });
const response = await client.send(new InvokeAgentCommand({
  agentId: "AGENT12345",
  agentAliasId: "ALIAS12345",
  sessionId: "my-session-1",
  inputText: "Hello agent",
}));

let text = "";
for await (const event of response.completion) {
  if (event.chunk?.bytes) text += new TextDecoder().decode(event.chunk.bytes);
}
console.log(text);
```

Same call against a local mock server (`@aws-sdk/client-bedrock-agent-runtime` 3.1138.0, Node 22), plus the error when a required field is missing:

```
server: POST /agents/AGENT12345/agentAliases/ALIAS12345/sessions/my-session-1/text
server: request body = {"inputText":"Hello agent"}
client: contentType=application/json sessionId=my-session-1
client: completion text = "Hello from the agent"

# same call without agentId
client: Error: No value provided for input HTTP label: agentId.
```

<details>
<summary>Where this comes from</summary>

- Required vs optional and HTTP binding: the service model [`bedrock-agent-runtime.json`](https://github.com/aws/aws-sdk-js-v3/blob/c68e50e4a6e0469a20c2894fe8a29c140553ebb8/codegen/sdk-codegen/aws-models/bedrock-agent-runtime.json) at main (c68e50e) marks `agentId`, `agentAliasId` and `sessionId` with `smithy.api#required` + `smithy.api#httpLabel`; `inputText` and the other members are body fields without `required`. The model as of July 2024 (f2f40c1) had the same three required. The generated type mirrors it: [`models_1.ts:2226-2297`](https://github.com/aws/aws-sdk-js-v3/blob/c68e50e4a6e0469a20c2894fe8a29c140553ebb8/clients/client-bedrock-agent-runtime/src/models/models_1.ts#L2226-L2297) has `agentId: string | undefined` next to `inputText?: string | undefined`.
- Response shape: `completion` is `AsyncIterable<ResponseStream>` ([`models_0.ts:7743`](https://github.com/aws/aws-sdk-js-v3/blob/c68e50e4a6e0469a20c2894fe8a29c140553ebb8/clients/client-bedrock-agent-runtime/src/models/models_0.ts#L7743)), each event a union with `chunk.bytes` ([`InvokeAgentCommand.ts:276-279`](https://github.com/aws/aws-sdk-js-v3/blob/c68e50e4a6e0469a20c2894fe8a29c140553ebb8/clients/client-bedrock-agent-runtime/src/commands/InvokeAgentCommand.ts#L276-L279)). The operation doc in that file says the final answer arrives in `chunk.bytes` and that `enableTrace: true` adds `trace` events.
- Full field list with types: [InvokeAgentCommand API reference](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/client/bedrock-agent-runtime/command/InvokeAgentCommand/).
- Since you asked, the operation doc gained a note that Bedrock Agents (now "Agents Classic") is [no longer open to new customers](https://docs.aws.amazon.com/bedrock/latest/userguide/agents-classic-maintenance-mode.html); existing agents keep working with this client.

<details>
<summary>Mock server script (no AWS account needed)</summary>

```js
// Discussion #6289: which parameters InvokeAgentCommand needs, and how the response comes back.
// Runs against a local mock HTTP/2 server so no AWS account is needed. MISSING=1 -> leave out agentId.
import http2 from "node:http2";
import { BedrockAgentRuntimeClient, InvokeAgentCommand } from "@aws-sdk/client-bedrock-agent-runtime";
import { EventStreamCodec } from "@smithy/eventstream-codec";
import { fromUtf8, toUtf8 } from "@smithy/util-utf8";

const codec = new EventStreamCodec(toUtf8, fromUtf8);
const chunk = (text) => Buffer.from(codec.encode({
  headers: {
    ":message-type": { type: "string", value: "event" },
    ":event-type": { type: "string", value: "chunk" },
    ":content-type": { type: "string", value: "application/json" },
  },
  body: fromUtf8(JSON.stringify({ bytes: Buffer.from(text).toString("base64") })),
}));

const server = http2.createServer();
server.on("stream", (stream, headers) => {
  let body = "";
  stream.setEncoding("utf8");
  stream.on("data", (d) => (body += d));
  stream.on("end", () => {
    console.log(`server: ${headers[":method"]} ${headers[":path"]}`);
    console.log(`server: request body = ${body}`);
    stream.respond({ ":status": 200, "content-type": "application/vnd.amazon.eventstream", "x-amzn-bedrock-agent-content-type": "application/json", "x-amz-bedrock-agent-session-id": "my-session-1" });
    for (const t of ["Hello", " from", " the agent"]) stream.write(chunk(t));
    stream.end();
  });
});
await new Promise((r) => server.listen(0, "127.0.0.1", r));

const client = new BedrockAgentRuntimeClient({
  region: "us-east-1",
  endpoint: `http://127.0.0.1:${server.address().port}`,      // remove this line to talk to the real service
  credentials: { accessKeyId: "AKIAEXAMPLE", secretAccessKey: "example" }, // and this one, to use the default credential chain
});

const params = {
  agentId: "AGENT12345",        // required (path)
  agentAliasId: "ALIAS12345",   // required (path)
  sessionId: "my-session-1",    // required (path) - any string you choose; reuse it to continue the conversation
  inputText: "Hello agent",     // the user's message (body)
};
if (process.env.MISSING === "1") delete params.agentId;

try {
  const response = await client.send(new InvokeAgentCommand(params));
  console.log(`client: contentType=${response.contentType} sessionId=${response.sessionId}`);
  let text = "";
  for await (const event of response.completion) {
    if (event.chunk?.bytes) text += new TextDecoder().decode(event.chunk.bytes);
  }
  console.log(`client: completion text = ${JSON.stringify(text)}`);
} catch (e) {
  console.log(`client: ${e.name}: ${e.message}`);
}
server.close();
```
</details>
</details>
